### ISSUE TITLE
Question about metrics

### ISSUE THREAD

**Issue by ahariri13** [huang-xx/STGAT#3]
Hi ! I just have a small confusion regarding the metric ADE. 

If I trained one datasets containing of the shape (200 000, 40, 5,2) where I have 40 frames per video, with ach video containing 5 walking people. I observe 20 frames and try to predict the next 20 frames for these 5 persons. I get ADE = 11 and FDE = 15.   

Is my Average Displacement Error calculated per frame per person ? Or should I divide by the number of frames (or persons ?)

**Comment by huang-xx** [comment]
@ahariri13   The Average Displacement Error is calculated per frame per person. You should get the total displacement errors of 5 pedestrians in 20 prediction frames, then divide (5*20).

**Comment by ahariri13** [comment]
@huang-xx Thank you for your reply. What I need is to obtain the ADE at each time step of the 20 predicted frames. I believe it should be doable within the evaluate function below in evaluate_model.py , but I was confused with the way ADE is calculated for batch of trajectories. Any help would be appreciated on where to append the time-step-specific ADE values would be appreciated :)  


**def evaluate(args, loader, generator):
     ade_outer, fde_outer = [], []
     total_traj = 0
     with torch.no_grad():
        for batch in loader:
            batch = [tensor.cuda() for tensor in batch]
            (
                obs_traj,
                pred_traj_gt,
                obs_traj_rel,
                pred_traj_gt_rel,
                non_linear_ped,
                loss_mask,
                seq_start_end,
            ) = batch
            ade, fde = [], []
            total_traj += pred_traj_gt.size(1)

            for _ in range(args.num_samples):
                pred_traj_fake_rel = generator(
                    obs_traj_rel, obs_traj, seq_start_end, 0, 3
                )
                pred_traj_fake_rel = pred_traj_fake_rel[-args.pred_len :]

                pred_traj_fake = relative_to_abs(pred_traj_fake_rel, obs_traj[-1])
                ade_, fde_ = cal_ade_fde(pred_traj_gt, pred_traj_fake)
                ade.append(ade_)
                fde.append(fde_)
            ade_sum = evaluate_helper(ade, seq_start_end)
            fde_sum = evaluate_helper(fde, seq_start_end)

            ade_outer.append(ade_sum)
            fde_outer.append(fde_sum)

        ade = sum(ade_outer) / (total_traj * args.pred_len)
        fde = sum(fde_outer) / (total_traj)
        return ade, fde**

Also, you defined ADE as being  the mean square error (MSE) over all estimated positions in the pre-dicted trajectory and ground-truth trajectory, but in utils.py the following was performed: 

    **loss = loss ** 2
    if consider_ped is not None:
        loss = torch.sqrt(loss.sum(dim=2)).sum(dim=1) * consider_ped
    else:
        loss = torch.sqrt(loss.sum(dim=2)).sum(dim=1)**

Isn't this function computing the Root Mean Square Error (RMSE) ? 

Best regards !

**Comment by huang-xx** [comment]
@ahariri13 Thank you for your correction, our statement about ADE in the paper is wrong, and it should be RMSE or L2 distance(as in [SocialAttention](https://arxiv.org/pdf/1710.04689.pdf) and [SocialGan](https://arxiv.org/abs/1803.10892)). Very sorry about that.

**Comment by ahariri13** [comment]
@huang-xx Thank you for clarifying :) any thought on how to obtain the ADE at each time step of the 20 predicted frames ? As I mentioned in the first part there should be a place in the evaluate function where I can append it, and getting those are urgent for a project I am working on. 

Thanks in advance !

**Comment by huang-xx** [comment]
@ahariri13 If I didn't misunderstand your question, you need to modify the ‘pred_len’ parameter from 8 to 20.

**Comment by ahariri13** [comment]
Well yes I did that but what I meant is: when I run evaluate_model.py, I get one ADE value (around 11 ) which is the  average of all ades on the 20 predicted frames. I would like to get Ade for predicted frame 1, predicted frame 2 ,..., predicted frame 20. 

Hope that was clear, sorry for the confusion. Thanks in advance :)

**Comment by huang-xx** [comment]
@ahariri13 You should modify the function 'cal_ade_fde()' in evaluate_model.py, and function 'displacement_error()' in utils.py. Specifically, Add the parameter 'seq_start_end' to the function parameter list. The parameter 'seq_start_end' contains grouping information ( indicate the splitting of pedestrians in the batch belong to the same frame ). You can get the average displacement error of each pedestrian in each frame by modifying the function 'displacement_error()'.
